src/extract_gpuinfo_intel_xe.c

```diff
@@ -0,0 +1,264 @@
+  struct gpuinfo_dynamic_info *dynamic_info = &gpu_info->base.dynamic_info;
+
+  if (gpu_info->card_fd) {
+    int32_t length = 0;
+    struct drm_xe_query_mem_regions *regions =
+        xe_device_query_alloc_fetch(gpu_info->card_fd, DRM_XE_DEVICE_QUERY_MEM_REGIONS, &length);
+    if (regions) {
+      for (unsigned i = 0; i < regions->num_mem_regions; i++) {
+        struct drm_xe_mem_region mr = regions->mem_regions[i];
+        if (mr.mem_class == DRM_XE_MEM_REGION_CLASS_VRAM || regions->num_mem_regions == 1) {
+          SET_GPUINFO_DYNAMIC(dynamic_info, total_memory, mr.total_size);
+          // xe will report 0 kb used if we don't have CAP_PERFMON
+          if (mr.used != 0) {
+            SET_GPUINFO_DYNAMIC(dynamic_info, used_memory, mr.used);
+            SET_GPUINFO_DYNAMIC(dynamic_info, free_memory, dynamic_info->total_memory - dynamic_info->used_memory);
+            SET_GPUINFO_DYNAMIC(dynamic_info, mem_util_rate, dynamic_info->used_memory * 100 / dynamic_info->total_memory);
+          }
+          break;
+        }
+      }
+      free(regions);
+    }
+  }
+}
+
+static const char xe_drm_intel_vram[] = "drm-total-vram0";
+static const char xe_drm_intel_gtt[] = "drm-total-gtt";
+// Render
+static const char xe_drm_intel_cycles_rcs[] = "drm-cycles-rcs";
+static const char xe_drm_intel_total_cycles_rcs[] = "drm-total-cycles-rcs";
+// Video Decode
+static const char xe_drm_intel_cycles_vcs[] = "drm-cycles-vcs";
+static const char xe_drm_intel_total_cycles_vcs[] = "drm-total-cycles-vcs";
+// Video Enhance
+static const char xe_drm_intel_cycles_vecs[] = "drm-cycles-vecs";
+static const char xe_drm_intel_total_cycles_vecs[] = "drm-total-cycles-vecs";
+// Copy
+static const char xe_drm_intel_cycles_bcs[] = "drm-cycles-bcs";
+static const char xe_drm_intel_total_cycles_bcs[] = "drm-total-cycles-bcs";
+// Compute
+static const char xe_drm_intel_cycles_ccs[] = "drm-cycles-ccs";
+static const char xe_drm_intel_total_cycles_ccs[] = "drm-total-cycles-ccs";
+
+static const char *cycles_keys[] = {xe_drm_intel_cycles_rcs, xe_drm_intel_cycles_vcs, xe_drm_intel_cycles_vecs,
+                                    xe_drm_intel_cycles_bcs, xe_drm_intel_cycles_ccs};
+
+static const char *total_cycles_keys[] = {xe_drm_intel_total_cycles_rcs, xe_drm_intel_total_cycles_vcs,
+                                          xe_drm_intel_total_cycles_vecs, xe_drm_intel_total_cycles_bcs,
+                                          xe_drm_intel_total_cycles_ccs};
+
+bool parse_drm_fdinfo_intel_xe(struct gpu_info *info, FILE *fdinfo_file, struct gpu_process *process_info) {
+  struct gpu_info_intel *gpu_info = container_of(info, struct gpu_info_intel, base);
+  static char *line = NULL;
+  static size_t line_buf_size = 0;
+  ssize_t count = 0;
+
+  bool client_id_set = false;
+  unsigned cid;
+  nvtop_time current_time;
+  nvtop_get_current_time(&current_time);
+
+  union intel_cycles gpu_cycles = {.array = {0, 0, 0, 0, 0}};
```

Comment:
If you want to zero-initialize you can also use `= { 0 }` ( or `= {}` if the project uses the C standard >= C23)

Reply:
Ahh cool, thank you!